src/llama.cpp

```diff
@@ -10243,11 +10280,31 @@ static struct ggml_tensor * llm_build_moe_ffn(
     ggml_tensor * logits = llm_build_lora_mm(lctx, ctx, gate_inp, cur); // [n_expert, n_tokens]
     cb(logits, "ffn_moe_logits", il);
 
-    ggml_tensor * probs = ggml_soft_max(ctx, logits); // [n_expert, n_tokens]
-    cb(probs, "ffn_moe_probs", il);
+    ggml_tensor * probs = nullptr;
+    switch (gating_op) {
+        case LLM_EXPERT_GATING_FUNC_SOFTMAX:
+            {
+                probs = ggml_soft_max(ctx, logits); // [n_expert, n_tokens]
+                cb(probs, "ffn_moe_probs", il);
+            } break;
+        case LLM_EXPERT_GATING_FUNC_SIGMOID:
+            {
+                probs = ggml_sigmoid(ctx, logits); // [n_expert, n_tokens]
+                cb(probs, "ffn_moe_sigm", il);
+            } break;
+        default:
+            GGML_ABORT("fatal error");
+    }
+
+    // add experts selection bias - introduced in DeepSeek V3
+    ggml_tensor * selection_probs = probs;
+    if (expert_weights_b != nullptr) {
+        selection_probs = ggml_add(ctx, probs, expert_weights_b);
+        cb(selection_probs, "ffn_moe_sigm_biased", il);
+    }
```

Comment:
Can be simplified to:

```suggestion
    // add experts selection bias - introduced in DeepSeek V3
    if (expert_weights_b != nullptr) {
        probs = ggml_add(ctx, probs, expert_weights_b);
        cb(probs, "ffn_moe_sigm_b", il);
    }
```

Reply:
I'm afraid this won't work correctly, as the original unmodified weights are still needed for multiplication with the experts output at the end of the function. Biased weights are used only for expert selection. See the [DeepSeek V3 technical report](https://github.com/deepseek-ai/DeepSeek-V3/blob/main/DeepSeek_V3.pdf): 

> Note that the bias term is only used for routing. The gating value, which will be multiplied with
> the FFN output, is still derived from the original affinity score

Reply:
I see - I missed this.